### PR TITLE
Adding Name Encoding Improvement without MbString Extension

### PR DESCRIPTION
**Pull request by SirLouen**
More information about this here:
https://github.com/PHPMailer/PHPMailer/issues/3201

The problem is that the section of the code that is decoding those values is not tackling well charsets, specially UTF-8. There might be more scenarios untackled, as handling this without `mbstring` feels a bit arbitrary. Feel free to add more cases in a review.

### CODE REVIEW

File: src/PHPMailer.php [hunk 1]
```diff
@@ -1321,6 +1301,69 @@ public static function parseAddresses($addrstr, $useimap = true, $charset = self
...
+            $prepared = str_replace('_', '=20', $value);
+            $decoded = mb_decode_mimeheader($prepared);
+            mb_internal_encoding($origCharset);
+
+            return trim($decoded, "'\" ");
+        }
+
+        // If no mbstring, use a manual decoder
+        $decoded = preg_replace_callback(
+            '/=\?([^?]+)\?([bqBQ])\?([^?]+)\?=/s',
```
[the remark this answers is not in the document]
Comment (SirLouen) [reply src/PHPMailer.php:1334]: @Synchro I'm not really sure what you mean. 
If you want to use multiple encodings at the same time, you might be using UTF-8 or even UTF-16 depending on which characters you want to use combined.

**Comment by SirLouen** [comment]
@Synchro I'm thinking now that the selected name for the function is unfortunate (although we could consider extending it). At this moment, it's only being targeted by the `parseAddresses` function. And more specifically, the name within an address (`The Name <the@address.com>`) 

With Imap parsing the `$address->personal`
and without IMAP parsing: `$name`

But it could definitely be 
A) extended to cover each single Header value in the system. (I have not reviewed the other, though)
B) Maybe better changing the name to `decodeAddressNameValue`

Also I've been thinking about the tests for no-mbstring

Why the expected value is the encoded message?
```
    'name'    => '=?utf-8?B?0J3QsNC30LLQsNC90LjQtSDRgtC10YHRgtCw?=',
                            'address' => 'encoded@example.org',
```

I modified this section https://github.com/SirLouen/PHPMailer/blob/bf8e97d3c83247948ca5537585224ae8a3ab3140/test/SendTestCase.php#L68-L75
```diff
            switch ($requestKey) {
                case 'mail_to':
                    $this->propertyChanges[$phpmailerKey] = [
                        'address' => $_REQUEST[$requestKey],
-                        'name'    => 'Test User',
+                        'name'    => 'Test ⭐ User',
                    ];
                    break;
```

And this is what I'm receiving in the environment without mbstring
Test â­ User <to@example.com>

And this in the regular environment
Test ⭐ User <to@example.com>

Although my tests are passing (I even added an extra test with emojis to the `dataAddressSplitting` dataset like:

```
                        [
                            'name'    => '=?UTF-8?B?SGVsbG8g8J+MjSDkuJbnlYwgY2Fmw6k=?=',
                            'address' => 'encoded2@example.org',
                        ],
```

Conclusion: The no-mbstring tests seem to be screwed… because the code for this use-case/environment is not right, if you dont have `mbstring` installed, any name not plain text will not be shown correctly

Worth the effort fixing this? As I said in 2025, I feel that `mbstring` is not an option any more, it feels very PHP 5.X.

Reading the original code, and my code, you can really see that what is doing is leaving most of the charset as-is (like UTF-8) and simply trying to do an effort to decode certain encodings like ISO-8859-1. I think that bulding a full decoder with all the charset that exist, etc... will be an arduous task not worthy without a lib, or an extesion like `mbstring`

At best, I would propose to limit certain charsets and force people without `mbstring` to restrict their `FromNames` to plaintext (anyway I doubt that anyone in this situation could have tried to use anything not plain text or at worst ISO-8859-1 that is the only supported charset atm, otherwise they might have reported because this has been like this for plenty of years)

**Comment by Synchro** [comment]
Running without `mbstring` is not a primary aim, and support does not have to be comprehensive. Historically PHPMailer has defaulted/assumed ISO-8859-1 (as has PHP), and that's something that can be handled without mbstring, so long as you don't expect any kind of clever charset mapping. There's no point in trying to implement an encoding engine when installing mbstring solves the problem so easily. I think it's fine for it to fail in ugly ways if you provide UTF-8 content when mbstring isn't available; just *don't do that*™. While not every ISO-8859-x string is valid UTF-8, the reverse is always true; it will look bad, but that's the caller's problem.

In short, we should not try to make UTF-8 decoding work without mbstring.

**Comment by SirLouen** [comment]
OK, I got it. I think I have all the info now. I'm going to sort a couple of troubles in the code with this edge case and refactor the unit tests in `ParseAddressesTest` to be more targeted. 

Also, the inline docs will reflect this for future reference.

**Comment by SirLouen** [comment]
@Synchro I needed to do a step backwards to first fix #3204 

I split the decoding code from the `parseAddresses` function so it could be easier to target the logic from the tests.

In the tests, now we could target on one side, the address splitting, and on the other side the decoding

No need to have a million combinations depending on the extensions enabled. 

Theoretically the splitting part should work identically for both `imap_rfc822_parse_adrlist` and “native” versions, so need to have two separate logics in the tests. So they should also both pass identically.

But on the decoding part things could be different. Without `mbstring`, you have said that we should be aiming for a downgraded version, right?.

**The big problem I'm finding now is that I don't really know what we are expecting from the downgraded version.** 

I think first we need to figure out and set the expectations for the tests with this "native" version. I provided a good amount of examples in the `dataDecodeHeader` data provider.

Also, why not using the `iconv_mime_decode`  in case the `mbstring` one is missing? Contrarily to the `mbstring` one, this is not an extension, and `iconv` must be disabled explicitely for this to not work, which is completely unlikely in any installation nowadays. In fact, I believe that the `iconv` can be better in many cases than the `mbstring` because it has a better built-in  non-compliant RFC2047 strings detector.

PS: To the question if `mbstring` was required in WordPress, [its not](
https://github.com/SirLouen/wordpress-develop/blob/94cab031c85eaf86e5e6975421359b0be3b994b7/src/wp-admin/includes/class-wp-site-health.php#L936-L940)

**Comment by SirLouen** [comment]
After thinking a bit more, I realized that PHPUnit tests are running by default in UTF-8. It makes no sense to selectively pass the encoding in the tests. 

About the `mbstring` distinction, as I said, `iconv_mime_decode` pretty much does the same, but better. It's possible to simply wipe `mb_decode_mimeheader` as it's not providing any additional value, and at least for this particular case, no distinction will be needed between `mbstring` and `no-mbstring` environments. 

I've revamped the whole tests and now looks cleaner and much more comprehensive overall, the previous version was very convoluted with a ton of uses cases that, as I commented in my previous messages, weren't really needed. 

A little last improvement could simply be, removing the `mbstring` functions part, and leaving just the `iconv` to handle plain everything. In fact, there was no handling at all for the previous version (the supposedly `no-mbstring` scenario), and strings were simply being passed-through without any decoding. This will do the same, but with an extra fallback. It's also possible to leave it, like: "if there is `mbstring` do it that way, if, not, try to use `iconv`, if neither of the two, which is extremely unlikely in 2025, then simply throw the string without any decoding). It convolutes the code, but I think it's well documented inline.

Finally, this is how are the tests looking now (including all [my PR group excluding](https://github.com/PHPMailer/PHPMailer/pull/3197))
image

And this for the `nombstring` version (still need to look into that `testHeaderEncoding` in `PHPMailerTest.php` More info: #3206)

image

Finally, I forgot, I removed the IMAP switch in the `parseAddresses` function. With a IMAP function availability check, it doesn't make any sense to disable it. I've tried to dig into the commit that introduce it, and there are no reasons for adding this flag. Using this function if exists is basically the most ideal. Maybe the reason is that when it was introduced it was experimental and could lead to errors, but nowadays not using it, if available, lost all sense (all to keep things cleaner, including tests)

**Comment by Synchro** [comment]
This is all good, and thanks for your efforts.

I didn't realise that iconv was a default extension, so it makes sense to use it if it's there, though I don't want to get into the situation of duplicating everything that mbstring does using iconv instead. We should add `ext-iconv` to the `composer.json` suggest section too. I've also seen that the imap extension is somewhat unloved, and often omitted from packaged PHP versions, hence the appearance of [imapengine](https://packagist.org/packages/directorytree/imapengine).

Regarding the fallback case, I don't think we need to care: if you want to use UTF-8, you must have mbstring. If you try to use it without, it's going to break.

On the `useimap` param, I agree; I don't quite recall why I added it at the time; the main thing was that I wanted the address parsing to be a standalone feature. If the automated choice is reliable, leave the param in place, mark it as deprecated, and ignore it internally. I deprecated `SingleTo` a while ago; that kind of thing should be done in the app layer, outside PHPMailer.

Separately, could you please sign your commits? Downstream packagers like that.

**Comment by SirLouen** [comment]
Ok @Synchro added the signed commit, added the deprecation and the composer suggestion.

**Comment by SirLouen** [comment]
Unit tests are failing because they are taking the old format of the unit tests, which use the no `imap` version, which happens to fail in a specific case that we need to fix in another PR. [This is related](https://github.com/PHPMailer/PHPMailer/issues/3210) and also mentions this failing test. But I would recommend approaching this separately to avoid a massive commit with a ton of fixes at the same time. 
Better to first merge this PR and then go for that.

**Comment by Synchro** [comment]
It's also failing because of [a bug in the composer-install GitHub action](https://github.com/ramsey/composer-install/issues/270) that breaks on PHP 8.5.

**Comment by Synchro** [comment]
So near... Now we have [another PHP 8.5 problem](https://github.com/PHPMailer/PHPMailer/actions/runs/17121380830/job/48629457012?pr=3202) triggered by the linter.

@jrfnl Do you know how we suppress that, since it's intentional?

**Comment by jrfnl** [comment]
> So near... Now we have [another PHP 8.5 problem](https://github.com/PHPMailer/PHPMailer/actions/runs/17121380830/job/48629457012?pr=3202) triggered by the linter.
> 
> @jrfnl Do you know how we suppress that, since it's intentional?

Yup, if it's intentional it can be ignored.

There are two/three options:
* Either ignore this file by default for the linter and have a separate lint run specifically for this file without the `--show-deprecated` flag.
* Add [a comment](https://github.com/php-parallel-lint/PHP-Parallel-Lint#excluding-files-from-a-scan-based-on-the-php-version-used) on the line containing the `<?php` tag to only lint this file on PHP < 8 .5: `<?php // lint < 8.5`.

However, as this is a security related test fixture, I would recommend splitting the test fixture file into two files - one containing everything except that one line, the other containing just that one line.

Then for the split off test fixture file with just the `ls-l` code, apply one of the above mentioned options.

As a last alternative, the backticks could be changed to `shell_exec()`.

If you let me know what you prefer, I'd be happy to send you a PR.

**Comment by Synchro** [comment]
Yes please to the PR. I think the one-line split is the best approach, and there is already a function call test in there, so we need to retain the backticks for < 8.5.

**Comment by jrfnl** [comment]
Regarding `iconv`, yes, it is a [default extension](https://www.php.net/manual/en/iconv.requirements.php), but only on recent POSIX compliant systems, and it is also an extension which [can be disabled](https://www.php.net/manual/en/iconv.installation.php).

If I remember correctly from some extension checks done by someone of the WP Core team quite a while back, Mbstring was in practice installed on more systems than iconv, but might be worth asking again to get some up-to-date numbers. (I believe Dion should be able to provide those)

**Comment by jrfnl** [comment]
> Yes please to the PR. I think the one-line split is the best approach, and there is already a function call test in there, so we need to retain the backticks for < 8.5.

See PR #3211

**Comment by Synchro** [comment]
OK, so PHP 8.5 linting is now OK, and we are down to the problem with the quotes in names that #3210 is about, so I'll wait for that fix before merging this one.

**Comment by Synchro** [comment]
Apologies,  I made a bit of a mess of the commit history while merging in changes from master.

**Comment by SirLouen** [comment]
We left this pending for an `mbstring`/`iconv` decission

I've been reviewing WP code and `iconv` is pretty much used as default it is (without checks for the most part). There is this `getID3` library for media parsing that uses `iconv` very intensively. 

A user in WP sent me this https://stackoverflow.com/questions/4050754/iconv-or-mbstring that preaches for a combination of the two for the most part.

But I asked in WP and [Paul ](https://profiles.wordpress.org/paulkevan/) told me that it seems that both `iconv` and `mbstring` have 99%+ use rate. So I feel that the `iconv` isn't not going to do any difference, and maybe we can introduce a user notice like we did with the `imap` library.

Also by default, `mbstring` is not removed in WP tests and most of the functions are not even protected, so despite is not a required option, it seems that it's treated as default. 

Conclusion: `E_USER_NOTICE` and no `iconv`

**Comment by SirLouen** [comment]
This is the answer by Dion

> Personally, having looked at the available stats…
> iconv is available on 0.4% more sites.  (99.3% vs 99.7%)
> If there’s an mbstring implementation already, I’d just stick with that.
> Chances are, if they have iconv they have mbstring.
> Because this primarily affects non-english sites.. let me query the numbers for English vs not-english… Maybe non-english sites have a different number…  Personally I expect any non-english host worth anything will have both available.

**Comment by Synchro** [comment]
So much mess! I look forward to making `ext-mbstring` a requirement in the the next x.0 release...

File: composer.json [hunk 2]
```diff
@@ -49,6 +49,7 @@
     },
     "suggest": {
         "decomplexity/SendOauth2": "Adapter for using XOAUTH2 authentication",
+        "ext-iconv": "To support UTF-8 if the mbstring PHP extension is not enabled",
```
Comment (Synchro) [inline composer.json:52]: This makes it sound like `iconv` is a complete replacement for `mbstring`, but I think we are only explicitly using it in `parseAddresses`, so that's misleading.

Comment (SirLouen) [reply composer.json:52]: I'm currently working on this. I will ping you out when I have this finished.

**Comment by SirLouen** [comment]
Ok, I think I got it now

1. Added the better parsing for #3235 
2. Better ordered the tests in `ParseAddressesTest.php`
3. Removed all traces of `iconv` as commented by the WP meta team `iconv` and `mbstring` are pretty much the same % so we can get rid of so much duplicated code. 
4. Now I can take advantage of the new static translator, so I can use this for the deprecated argument
5. And now, all methods have their specific test coverage counterparts.

**Comment by SirLouen** [comment]
Edited: I have switched to PHP 8.4 and I've been able to reproduce this. I will look these pesky extra spaces tomorrow.

**Comment by SirLouen** [comment]
After some debugging I've been able to find the culprit

In 8.3 they introduced some changes in `mb_decode_mimeheader`

> [mb_decode_mimeheader()](https://www.php.net/manual/en/function.mb-decode-mimeheader.php) interprets underscores in QPrint-encoded MIME encoded words as required by RFC 2047; they are converted to spaces. Underscores must be encoded as "=5F" in such MIME encoded words.

Which could be the culprit of this problem I'm finding now
https://3v4l.org/ZF35Z

I'm going to look for a solution for this

**Comment by SirLouen** [comment]
@Synchro finally all troubles have been sorted out. Now you can review this. 

These are essentially the changes
https://github.com/PHPMailer/PHPMailer/pull/3202#issuecomment-3336223245

+ Checking for those pesky spaces in 8.3+ in the `mbstring` parsing section.

File: src/PHPMailer.php [hunk 3]
```diff
@@ -1237,16 +1237,19 @@ protected function addAnAddress($kind, $address, $name = '')
...
-     * @param bool   $useimap Whether to use the IMAP extension to parse the list
-     * @param string $charset The charset to use when decoding the address list string.
+     * @param string      $addrstr       The address list string
+     * @param string|null $deprecatedArg Deprecated argument since 6.11.0.
+     * @param string      $charset       The charset to use when decoding the address list string.
      *
      * @return array
      */
-    public static function parseAddresses($addrstr, $useimap = true, $charset = self::CHARSET_ISO88591)
+    public static function parseAddresses($addrstr, $deprecatedArg = null, $charset = self::CHARSET_ISO88591)
```
[the remark this answers is not in the document]
Comment (SirLouen) [reply src/PHPMailer.php:1246]: I think that we could perfectly roll back this little change because now we have the right notice text, that we didn't before the "static-fication" of the `setLanguage` function. When I added the first code, it felt a little confusing of why this was being deprecated without adding some change like this, but now [it's clearly advertised](https://github.com/PHPMailer/PHPMailer/blob/6ed58a9c607f1830bea0e58f88cf30f7344f8b3f/src/PHPMailer.php#L1253) if anyone tries to use this argument without using a `null`. 

What I'm not 100% clear, maybe @jrfnl knows more about this, when an expected value is no more the value in particular, in this case `bool` shall we maintain the docs typing? Docs won't provoke any errors, but they will alert the developers in their IDE (apart from the obvious runtime notice) that they should be changing their code (if they are using a bool, they will get the IDE error like: `bool` is not admitted. 

> Where does the string come from ? And shouldn't it allow for bool in case people still pass the old argument ?
Yes, the `string` is my mistake, when I was researching for a deprecation reference to apply here, I copied this from somewhere I can't remember, and I completely forgot to remove the `string` type. 

Conclusion: 
I would get back the same naming for BC, remove the `string` typing, and only leave `null` as the sole accepted value.

Makes sense @jrfnl @Synchro?
